### PR TITLE
[CARBONDATA-2023][DataLoad] Add size base block allocation in data loading

### PR DESCRIPTION
**Pull request by xuchuanyin**
Carbondata assign blocks to nodes at the beginning of data loading.
Previous block allocation strategy is block number based and it will
suffer skewed data problem if the size of input files differs a lot.

We introduced a size based block allocation strategy to optimize data
loading performance in skewed data scenario.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 `Only changed the internal interfaces`
 - [x] Any backward compatibility impacted?
 `No`
 - [x] Document update required?
`Updated the document`
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
`Added tests to verify the block-allocation correctness`
        - How it is tested? Please attach test report.
`Tested in local 3-node cluster`
        - Is it a performance related change? Please attach the performance test report.
```
In my scenario, the size of input data file varies from 1KB to about 5GB. 
Before enabling this feature, each executor processed the same number of blocks
 and the processed data size had a 5X gap.  --(block number based allocation)
After enabling this feature, each executor processed almost the same size of data
 and the processed number of data blocks had 6X gap. -- (block size based allocation)

The data loading performance had been promoted from 41MB/s/Node to 61MB/s/Node, 
about 50% performance enhancement gained.
```

        - Any additional information to help reviewers in testing this change.
       `I refactored the code to make it more readable. The core code mainly lies in CarbonLoaderUtil`
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
`Not related`

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2901/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1575/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2811/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1581/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2816/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1588/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2824/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2909/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2911/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2849/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1620/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2935/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2862/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1629/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2944/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2945/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2891/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1660/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2894/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1677/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2910/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1696/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2929/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1702/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2959/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2936/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1713/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2947/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2952/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1723/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2977/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2984/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2970/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1740/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3246/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3336/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2100/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3369/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2133/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3577/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2339/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3427/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java [hunk 1]
```diff
@@ -114,4 +114,14 @@
...
 
+  /**
+   * enable block size based block allocation while loading data. By default, carbondata assigns
+   * blocks to node based on block number. If this option is set to `true`, carbondata will
+   * consider block size first and make sure that all the nodes will process almost equal size of
+   * data. This option is especially useful when you encounter skewed data.
+   */
+  @CarbonProperty
+  public static final String ENABLE_CARBON_LOAD_SKEWED_DATA_OPTIMIZATION
+      = "carbon.load.skewed.data.optimization";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java]: change to `carbon.load.skewedDataOptimization.enabled`

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3598/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2360/

File: core/src/main/java/org/apache/carbondata/core/datastore/block/TableBlockInfo.java [hunk 2]
```diff
@@ -420,4 +436,16 @@ public String getDataMapWriterPath() {
   public void setDataMapWriterPath(String dataMapWriterPath) {
     this.dataMapWriterPath = dataMapWriterPath;
   }
+
+  @Override public String toString() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/block/TableBlockInfo.java]: move @Override to previous line

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 3]
```diff
@@ -1184,6 +1184,17 @@ public String getSortTempCompressor() {
       return CarbonCommonConstants.CARBON_SORT_TEMP_COMPRESSOR_DEFAULT;
     }
   }
+
+  /**
+   * whether optimization for skewed data is enabled
+   * @return true, if enabled; false for not enabled.
+   */
+  public boolean isCarbonLoadSkewedDataOptimizationEnabled() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: change to `isLoadSkewedDataOptimizationEnabled`

File: docs/useful-tips-on-carbondata.md [hunk 4]
```diff
@@ -169,5 +169,6 @@
...
   | carbon.sort.temp.compressor | spark/carbonlib/carbon.properties | Data loading | Specify the name of compressor to compress the intermediate sort temporary files during sort procedure in data loading. | The optional values are 'SNAPPY','GZIP','BZIP2','LZ4' and empty. By default, empty means that Carbondata will not compress the sort temp files. This parameter will be useful if you encounter disk bottleneck. |
+  | carbon.load.skewed.data.optimization | spark/carbonlib/carbon.properties | Data loading | Whether to enable size based block allocation strategy for data loading. | Carbondata will use number based block allocation strategy by default and it will make sure that all the executors process the same number of blocks. If this value is set to true, Carbondata will make sure that all the executors process the same size of data -- It's useful if the size of your input data files varies widely, say 1MB~1GB. |
```
Comment (jackylk) [inline docs/useful-tips-on-carbondata.md]: `Carbondata will use number based block allocation strategy by default` change to 
`When loading, carbondata will use file size based block allocation strategy for task distribution`

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 5]
```diff
@@ -70,6 +79,22 @@
...
+   */
+  public enum BlockAssignmentStrategy {
+    BLOCK_NUM_FIRST("Assign blocks to node base on number of blocks"),
+    BLOCK_SIZE_FIRST("Assign blocks to node base on data size of blocks");
+    private String name;
+    BlockAssignmentStrategy(String name) {
+      this.name = name;
+    }
+
+    @Override public String toString() {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: move @Override to previous line, please follow this in future

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 6]
```diff
@@ -455,83 +481,60 @@ public static Dictionary getDictionary(AbsoluteTableIdentifier absoluteTableIden
...
-  }
-
   /**
    * This method will divide the blocks among the nodes as per the data locality
    *
    * @param blockInfos
    * @param noOfNodesInput -1 if number of nodes has to be decided
    *                       based on block location information
+   * @param blockAssignmentStrategy strategy used to assign blocks
    * @return
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: please complete the description for @return

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 7]
```diff
@@ -606,122 +609,296 @@ private static void createTaskListForNode(Map<String, List<List<Distributable>>>
   }
 
   /**
-   * If any left over data blocks are present then assign those to nodes in round robin way.
+   * If any left over data blocks are present then assign those to nodes in round robin way. This
+   * will not obey the data locality.
    *
    * @param outputMap
-   * @param uniqueBlocks
+   * @param leftOverBlocks
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: Please remove @param if no description

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 8]
```diff
@@ -606,122 +609,296 @@ private static void createTaskListForNode(Map<String, List<List<Distributable>>>
...
 
   /**
-   * The method populate the blockLst to be allocate to a specific node.
-   * @param uniqueBlocks
-   * @param noOfBlocksPerNode
+   * assign blocks to nodes
+   * @param remainingBlocks
+   * @param expectedSizePerNode
+   * @param blockLst
+   * @param blockAssignmentStrategy
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: Please remove @param if no description

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 9]
```diff
@@ -606,122 +609,296 @@ private static void createTaskListForNode(Map<String, List<List<Distributable>>>
...
+      case BLOCK_SIZE_FIRST:
+        populateBlocksBySize(remainingBlocks, expectedSizePerNode, blockLst);
+        break;
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported block assignment strategy: " + blockAssignmentStrategy);
+    }
+  }
+  /**
+   * allocate blocks by block num
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: Take N number of distributable blocks from `remainingBlocks` and add them to output parameter `blockList`. The number of blocks add accumulated size is less than `expectedSizePerNode`

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 10]
```diff
@@ -606,122 +609,296 @@ private static void createTaskListForNode(Map<String, List<List<Distributable>>>
...
       blocks.remove();
-      if (blockLst.size() >= noOfBlocksPerNode) {
+      if (blockLst.size() >= expectedSizePerNode) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * allocate blocks by block size
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: Change to:
Take N number of distributable blocks from remainingBlocks and add them to output parameter blockList. The number of blocks added is determined by accumulated block size is less than expectedSizePerNode

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 11]
```diff
@@ -606,122 +609,296 @@ private static void createTaskListForNode(Map<String, List<List<Distributable>>>
...
    * To create the final output of the Node and Data blocks
    *
-   * @param outputMap
-   * @param blocksPerNode
-   * @param uniqueBlocks
-   * @param nodeAndBlockMapping
+   * @param outputNode2Blocks
+   * @param expectedSizePerNode
+   * @param remainingBlocks
+   * @param inputNode2Blocks
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: move @param if no description

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3608/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2370/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3609/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3456/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2373/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3610/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3613/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3619/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2379/

**Comment by jackylk** [comment]
merged into carbonstore branch

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3656/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3494/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2417/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2446/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3686/

**Comment by xuchuanyin** [comment]
retest sdv please

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3718/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2478/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3721/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2481/

**Comment by jackylk** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3520/

**Comment by jackylk** [comment]
Merged to carbonstore branch

**Pull request closed without merging**